control-plane/cni/main.go

```diff
@@ -0,0 +1,223 @@
+	injected        = "injected"
+)
+
+type CNIArgs struct {
+	types.CommonArgs
+	IP                         net.IP
+	K8S_POD_NAME               types.UnmarshallableString
+	K8S_POD_NAMESPACE          types.UnmarshallableString
+	K8S_POD_INFRA_CONTAINER_ID types.UnmarshallableString
+}
+
+// PluginConf is whatever you expect your configuration json to be. This is whatever
+// is passed in on stdin. Your plugin may wish to expose its functionality via
+// runtime args, see CONVENTIONS.md in the CNI spec.
+type PluginConf struct {
+	// This embeds the standard NetConf structure which allows your plugin
+	// to more easily parse standard fields like Name, Type, CNIVersion,
+	// and PrevResult.
+	types.NetConf
+
+	RuntimeConfig *struct {
+		SampleConfig map[string]interface{} `json:"sample_config"`
+	} `json:"runtime_config"`
+
+	Name string `json:"name"`
+	// Type of plugin (consul-cni)
+	Type string `json:"type"`
+	// CNIBinDir is the location of the cni config files on the node. Can bet as a cli flag.
+	CNIBinDir string `json:"cni_bin_dir"`
+	// CNINetDir is the locaion of the cni plugin on the node. Can be set as a cli flag.
+	CNINetDir string `json:"cni_net_dir"`
+	// Multus is if the plugin is a multus plugin. Can be set as a cli flag.
+	Multus bool `json:"multus"`
+	// Kubeconfig file name. Can be set as a cli flag.
+	Kubeconfig string `json:"kubeconfig"`
+	// LogLevl is the logging level. Can be set as a cli flag.
+	LogLevel string `json:"log_level"`
+}
+
+// parseConfig parses the supplied configuration (and prevResult) from stdin.
+func parseConfig(stdin []byte) (*PluginConf, error) {
+	cfg := PluginConf{}
+
+	if err := json.Unmarshal(stdin, &cfg); err != nil {
+		return nil, fmt.Errorf("failed to parse network configuration: %v", err)
+	}
+
+	// Parse previous result. This will parse, validate, and place the
+	// previous result object into conf.PrevResult. If you need to modify
+	// or inspect the PrevResult you will need to convert it to a concrete
+	// versioned Result struct.
+	if err := version.ParsePrevResult(&cfg.NetConf); err != nil {
+		return nil, fmt.Errorf("could not parse prevResult: %v", err)
+	}
+	// End previous result parsing
+
+	// Do any validation here
+	// TODO: Do validation
+	//	if conf.AnotherAwesomeArg == "" {
+	//		return nil, fmt.Errorf("anotherAwesomeArg must be specified")
+	//	}
+
+	return &cfg, nil
+}
+
+// cmdAdd is called for ADD requests.
+func cmdAdd(args *skel.CmdArgs) error {
+	cfg, err := parseConfig(args.StdinData)
+	if err != nil {
+		return err
+	}
+
+	// Get the values of args passed through CNI_ARGS
+	cniArgs := CNIArgs{}
+	if err := types.LoadArgs(args.Args, &cniArgs); err != nil {
+		return err
+	}
+
+	podNamespace := string(cniArgs.K8S_POD_NAMESPACE)
+	podName := string(cniArgs.K8S_POD_NAME)
+
+	// We only run in a pod
+	if podNamespace == "" && podName == "" {
+		return fmt.Errorf("not running in a pod, namespace and pod should have values")
+	}
+
```

Comment:
This logging goes to stdout but that is stdout on the node itself. Calico logs the output to /var/log on the filesystem and we might want to do something similar. I am worried about log rotation and filling up the file system of the node though...

I have seen Multus get messages into the pod description and I wonder if it is possible to send logs to the pod logs. We do have access to the kube api.  I think it would be a neat little feature and would standout as a nice usability feature.